solr/core/src/java/org/apache/solr/storage/SizeAwareDirectory.java

```diff
@@ -284,23 +405,37 @@ private SizeAccountingIndexOutput(
       this.fileSizeMap = fileSizeMap;
     }
 
-    public long setSizeWriter(LongObjectProcedure<String> sizeWriter) {
+    public Sizes setSizeWriter(SizeWriter sizeWriter) {
       if (this.sizeWriter == sizeWriter) {
-        return 0;
+        return new Sizes(0, 0);
       } else {
         // NOTE: there's an unavoidable race condition between these two
         // lines, and this ordering may occasionally overestimate directory size.
         this.sizeWriter = sizeWriter;
-        return backing.getFilePointer();
+        long onDiskSize = this.getFilePointer();
+        if (backing instanceof CompressingDirectory.SizeReportingIndexOutput) {
+          onDiskSize = ((CompressingDirectory.SizeReportingIndexOutput) backing).getBytesWritten();
+        }
+        return new Sizes(this.getFilePointer(), onDiskSize);
       }
     }
 
     @Override
     @SuppressWarnings("try")
     public void close() throws IOException {
+      long size = backing.getFilePointer();
       try (backing) {
-        fileSizeMap.put(name, backing.getFilePointer());
+        // TODO
       } finally {
+        long onDiskSize = size;
+        if (backing instanceof CompressingDirectory.SizeReportingIndexOutput) {
+          long finalBytesWritten = getBytesWritten(backing);
+          onDiskSize = finalBytesWritten;
+          // logical size should already be set through writeByte(s), but we need to finalize the
+          // on-disk size here
+          sizeWriter.apply(0, finalBytesWritten - lastBytesWritten, name);
+        }
+        fileSizeMap.add(name, new Sizes(backing.getFilePointer(), onDiskSize));
```

Comment:
Yeah this is the same type of problem, and it will indeed lead to under-reporting for temp files. This will be invoked for temp files (which don't write to persistent storage. The code (correctly) won't update `sizeWriter` for `onDiskSize` for temp files, but here since it initializes `long onDiskSize = size`, then would for temp files skip the `instanceof` conditional, and write a default non-zero `onDiskSize` to `fileSizeMap`.

The problem comes upon [SizeAwareDirectory.deleteFile()](https://github.com/cowpaths/fullstory-solr/blob/cc32c888770770f24792a5731996bb3336d3e64a/solr/core/src/java/org/apache/solr/storage/SizeAwareDirectory.java#L346-L355), which will find a non-zero onDiskSize for the temp file to be deleted, and subtracts it via `sizeWriter`.